scylladb/scylladb.go

```diff
@@ -0,0 +1,170 @@
+package scylladb
+
+import (
+	"errors"
+	"fmt"
+	"github.com/gocql/gocql"
+	"time"
+)
+
+// Storage interface that is implemented by storage providers
+type Storage struct {
+	session   *gocql.Session
+	tableName string
+
+	selectQuery string
+	insertQuery string
+	deleteQuery string
+	resetQuery  string
+}
+
+var (
+	checkSchemaMsg = "the `value` row has an incorrect data type. " +
+		"The message should be BLOB, but it is instead %s. This could lead to encoding-related issues if the database is not migrated (refer to https://github.com/gofiber/storage/blob/main/MIGRATE.md)"
+	createKeyspaceQuery = "CREATE KEYSPACE IF NOT EXISTS %s WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};"
+	dropQuery           = `DROP TABLE IF EXISTS %s.%s;`
+	createTableQuery    = `CREATE TABLE IF NOT EXISTS %s.%s (key TEXT PRIMARY KEY, value BLOB)`
+	checkSchemaQuery    = `SELECT type FROM system_schema.columns WHERE keyspace_name = '%s' AND table_name = '%s' AND column_name = 'value';`
+	keyspaceMsg         = `Keyspace cannot be empty.`
+)
+
+// New creates a new storage
+func New(config ...Config) *Storage {
+	var err error
+	var session *gocql.Session
+
+	// Set default config
+	cfg := configDefault(config...)
+
+	if cfg.Keyspace == "" {
+		panic(keyspaceMsg)
+	}
+
+	if cfg.Session == nil {
+		// Create a new cluster
+		cluster := gocql.NewCluster(cfg.Hosts...)
+		cluster.Consistency = gocql.ParseConsistency(cfg.Consistency)
+		cluster.Port = cfg.Port
+
+		// Set credentials if provided
+		if cfg.Username != "" && cfg.Password != "" {
+			cluster.Authenticator = gocql.PasswordAuthenticator{
+				Username: cfg.Username,
+				Password: cfg.Password,
+			}
+		}
+
+		// Set cfg.SslOpts if provided.
+		if cfg.SslOpts != nil {
+			cluster.SslOpts = cfg.SslOpts
+		}
+
+		// Create session
+		session, err = cluster.CreateSession()
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		// Set session if provided
+		session = cfg.Session
+	}
+
+	// Create keyspace if it does not exist
+	if err = session.Query(fmt.Sprintf(createKeyspaceQuery, cfg.Keyspace)).Exec(); err != nil {
+		session.Close()
+		panic(err)
+	}
+
+	// Drop table if reset is true
+	if cfg.Reset {
+		if err = session.Query(fmt.Sprintf(dropQuery, cfg.Keyspace, cfg.Table)).Exec(); err != nil {
```

Comment:
Just curious. Do we get panic if bucket is empty?

Reply:
Yes, you will get the error message "gocql: no streams available on connection"
Maximum number of inflight requests allowed per connection defaults to 3276.